dali/kernels/common/find/find_first_last_gpu.cuh

```diff
@@ -0,0 +1,229 @@
+ * @tparam Idx Index data type (int64_t, int32_t)
+ * @tparam Predicate Predicate that must be satisfied
+ * @tparam OutFormat Optional Transformation to first, last coordinates
+ * @param samples Sample descriptors, including a per-sample predicate
+ * @param format Optional output coordinate transformation
+ */
+template <typename T, typename Idx, typename Predicate, typename OutFormat>
+__global__ void FindFirstLastImpl(SampleDesc<T, Idx, Predicate, OutFormat> *samples) {
+  // This kernel is an adapted version of other reduce kernels (see ReduceAllBatchedKernel)
+  // This kernel processes 1024-element blocks laid out as 32x32.
+  // Grid is flat 2D and blockIdx.x corresponds to an output bin (a single element in this case)
+  // and blockIdx.y corresponds to sample in the batch.
+  // First, each thread goes with x-grid-sized stride over the data and iteratively reduces
+  // in a local variable.
+  // Then the local variable is reduced over the block.
+  int sample_idx = blockIdx.y;
+  auto &sample = samples[sample_idx];
+  const T *input = sample.in;
+  int64_t sample_len = sample.len;
+  auto &predicate = sample.predicate;
+  auto &format = sample.format;
+
+  const int64_t blk_size = blockDim.x * blockDim.y;
+  const int64_t grid_size = gridDim.x * blk_size;
+  const int64_t flat_tid = threadIdx.x + threadIdx.y * blockDim.x;
+
+  reductions::min first_reduction;
+  reductions::max last_reduction;
+
+  Idx first_neutral = first_reduction.template neutral<Idx>();
+  Idx last_neutral = last_reduction.template neutral<Idx>();
+  Idx first = first_neutral;
+  Idx last = last_neutral;
+
+  // similar concept as in reduction kernels
+
+  int64_t idx = blockIdx.x * blk_size + flat_tid;
+  for (; idx < sample_len; idx += grid_size) {
+    Idx tmp_idx = predicate(input[idx]) ? idx : -1;
+    Idx first_candidate = tmp_idx < 0 ? first_reduction.template neutral<Idx>() : tmp_idx;
+    Idx last_candidate = tmp_idx < 0 ? last_reduction.template neutral<Idx>() : tmp_idx;
+
+    first_reduction(first, first_candidate);
+    last_reduction(last, last_candidate);
+  }
+
+  BlockReduce(first, first_reduction);
+  BlockReduce(last, last_reduction);
+  if (flat_tid == 0) {
+    auto tmp =
+        (first == first_neutral || last == last_neutral) ? format.neutral() : format({first, last});
+    *sample.a_ptr = tmp.a;
+    *sample.b_ptr = tmp.b;
+  }
+}
+
+}  // namespace
+
+/**
+ * @brief Extract the position of the first and last position that satisfies a predicate
+ */
+class FindFirstLastGPU {
+ public:
+  template <typename T>
+  KernelRequirements Setup(KernelContext &context, const InListGPU<T, 1> &in) {
+    KernelRequirements req;
+    int nsamples = in.size();
+    TensorListShape<0> out_sh(nsamples);
+    req.output_shapes = {out_sh, out_sh};  // begin, length outputs
+    return req;
+  }
+
+  /**
+   * @brief Run from sample descriptors directly in GPU memory
+   * @remarks Convenient if predicates are to be generated on GPU
```

Comment:
I think this comment is relevant to the whole `sample_descs_gpu`, not only predicate.